homeassistant/components/snooz/fan.py

```diff
@@ -0,0 +1,155 @@
+"""Fan representation of a Snooz device."""
+from __future__ import annotations
+
+from collections.abc import Callable
+import logging
+from typing import Any
+
+from pysnooz.api import SnoozDeviceState, UnknownSnoozState
+from pysnooz.commands import (
+    SnoozCommandData,
+    SnoozCommandResultStatus,
+    set_volume,
+    turn_off,
+    turn_on,
+)
+from pysnooz.device import SnoozConnectionStatus, SnoozDevice
+
+from homeassistant.components.fan import ATTR_PERCENTAGE, FanEntity, FanEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, STATE_OFF, STATE_ON
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.restore_state import RestoreEntity
+
+from .const import DOMAIN
+from .models import SnoozConfigurationData
+
+# transitions logging is pretty verbose, so only enable warnings/errors
+logging.getLogger("transitions.core").setLevel(logging.WARNING)
+
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.DEBUG)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Snooz device from a config entry."""
+
+    address: str = entry.data[CONF_ADDRESS]
+    data: SnoozConfigurationData = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities(
+        [
+            SnoozFan(
+                hass,
+                data.device.display_name,
+                address,
+                data.device,
+            )
+        ]
+    )
+
+
+class SnoozFan(FanEntity, RestoreEntity):
+    """Fan representation of a Snooz device."""
+
+    def __init__(self, hass, name: str, address: str, device: SnoozDevice) -> None:
+        """Initialize a Snooz fan entity."""
+        self.hass = hass
+        self._address = address
+        self._device = device
+        self._attr_unique_id = address
+        self._attr_supported_features = FanEntityFeature.SET_SPEED
+        self._attr_name = name
+        self._attr_is_on = None
+        self._attr_should_poll = False
+        self._attr_percentage = None
+
+    def _write_state_changed(self) -> None:
+        # cache state for restore entity
+        if not self.assumed_state:
+            self._attr_is_on = self._device.state.on
+            self._attr_percentage = self._device.state.volume
+
+        self.async_write_ha_state()
+
+    def _on_connection_status_changed(self, new_status: SnoozConnectionStatus) -> None:
+        self._write_state_changed()
+
+    def _on_device_state_changed(self, new_state: SnoozDeviceState) -> None:
+        self._write_state_changed()
```

Comment:
It looks like these could be lambdas instead since you don't care about the arg?

Reply:
We could, but pylint doesn't seem to like the lambda.

![image](https://user-images.githubusercontent.com/499591/193435205-a9aa26ca-ed82-45af-bca3-13305917ffb1.png)

Not sure if this is what you had in mind.

Reply:
```diff
diff --git a/homeassistant/components/snooz/fan.py b/homeassistant/components/snooz/fan.py
index 17b8046a92..3a567e4f33 100644
--- a/homeassistant/components/snooz/fan.py
+++ b/homeassistant/components/snooz/fan.py
@@ -75,12 +75,6 @@ class SnoozFan(FanEntity, RestoreEntity):
 
         self.async_write_ha_state()
 
-    def _on_connection_status_changed(self, new_status: SnoozConnectionStatus) -> None:
-        self._write_state_changed()
-
-    def _on_device_state_changed(self, new_state: SnoozDeviceState) -> None:
-        self._write_state_changed()
-
     async def async_added_to_hass(self) -> None:
         """Restore state and subscribe to device events."""
         await super().async_added_to_hass()
@@ -97,13 +91,15 @@ class SnoozFan(FanEntity, RestoreEntity):
     @callback
     def _async_subscribe_to_device_events(self) -> Callable[[], None]:
         events = self._device.events
+        _on_connection_status_changed = lambda _: self._write_state_changed()
+        _on_device_state_changed = lambda _: self._write_state_changed()
 
         def unsubscribe():
-            events.on_connection_status_change -= self._on_connection_status_changed
-            events.on_state_change -= self._on_device_state_changed
+            events.on_connection_status_change -= _on_connection_status_changed
+            events.on_state_change -= _on_device_state_changed
 
-        events.on_connection_status_change += self._on_connection_status_changed
-        events.on_state_change += self._on_device_state_changed
+        events.on_connection_status_change += _on_connection_status_changed
+        events.on_state_change += _on_device_state_changed
 
         return unsubscribe
 
```

You could write it like that but honestly its not any better.  It might be nicer if you have `events` subscriptions returning callables to cancel subscriptions so you don't have to track the references to the function in hass